scripts/igmp_query_injector.py

```diff
@@ -274,37 +160,36 @@ def inject(self):
             log.info('Inject IGMP query without connection state check')
             self._inject_without_connection_state_check()
 
-    def _inject_with_connection_state_check(self):
-        vifs = [Vif(vif) for vif in self.vifs]
-        expected_vif_state = '4'
-        watch = XSVifStateWatch(expected_vif_state)
-        for vif in vifs:
-            watch.watch(vif.state_path, vif)
-
-        t = threading.Thread(target=watch.process_all_watches)
-        t.daemon = True
-        t.start()
-        remain_time = self.vif_connected_timeout
-
-        while remain_time > 0:
-            start_time = time.time()
-            try:
-                task = watch.get_task(remain_time)
-                vif = task.vif
+    def check_and_inject(self, watcher):
+        while watcher.watches:
+            path, vif = watcher.read_watch()
+            if g_xs_handler.read('', path) == vif.expected_state:
                 self.inject_to_vif(vif.vif_name, vif.mac)
-                vif.mark_injected()
-                if task.is_last:
-                    return
-            except Queue.Empty as e:
-                # timeout
-                break
+                watcher.unwatch(path, vif)
 
-            remain_time = remain_time - (time.time() - start_time)
+    def _inject_with_connection_state_check(self):
+        watcher = XSWatcher()
+        for vif in self.vifs:
+            vif = Vif(vif)
+            watcher.watch(vif.state_path, vif)
+
+        # We have 2 options to handle the blocking function `xs.read_watch`
+        # 1. Single thread with Unix alarm signal
+        #    The signal handler cannot be properly invoked because when signal received the process context is in
+        #    C extension but not python interpreter, so the interpreter cannot call the signal handler.
+        #    So we should apply this option.
+        # 2. Multi thread with threading.event or timeout associated join
+        #    Create a new thread to invoke the blocking methon `xs.read_watch`. In the main thread using a timeout
+        #    associated join method for waiting. We will apply this option.
+        t = threading.Thread(target=self.check_and_inject, args=(watcher,))
+        t.daemon = True
 
-        for vif in vifs:
-            if not vif.injected:
-                log.warning("Value of '%s' does not change to '%s' in %d seconds." %
-                            (vif.state_path, expected_vif_state, self.vif_connected_timeout))
+        t.start()
+        t.join(self.vif_connected_timeout)
```

Comment:
So you did not parallelize the whole process, in that if one vif hangs, all the other vifs has to wait?

Reply:
The parallelize stuff is taken by `xs.read_watch`, it works like `select`, so no vif which is ready will be blocked by other vifs.

Reply:
it is true for vif or vifs, but for bridges, it blocks.

Reply:
for bridge won't go into this branch